grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/sets/ShortcutFragmentSet.java

```diff
@@ -19,24 +19,126 @@
-                Fragments.shortcut(relationType, roleTypeA, roleTypeB, rolePlayerA, rolePlayerB),
-                Fragments.shortcut(relationType, roleTypeB, roleTypeA, rolePlayerB, rolePlayerA)
+                Fragments.inShortcut(rolePlayer, edge, relation, roleType, relationType),
+                Fragments.outShortcut(relation, edge, rolePlayer, roleType, relationType)
         );
     }
+
+    static boolean applyShortcutOptimisation(Collection<EquivalentFragmentSet> fragmentSets, GraknGraph graph) {
+        Iterable<CastingFragmentSet> castingFragmentSets =
+                fragmentSetOfType(CastingFragmentSet.class, fragmentSets)::iterator;
+
+        for (CastingFragmentSet castingFragmentSet : castingFragmentSets) {
+            if (attemptOptimiseCasting(fragmentSets, graph, castingFragmentSet)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private static boolean attemptOptimiseCasting(Collection<EquivalentFragmentSet> fragmentSets, GraknGraph graph, CastingFragmentSet castingFragmentSet) {
+        VarName relation = castingFragmentSet.relation();
+        VarName casting = castingFragmentSet.casting();
+
+        RolePlayerFragmentSet rolePlayerFragmentSet = findRolePlayerFragmentSet(fragmentSets, casting);
+
+        // Try and get type of relation
+        Optional<IsaFragmentSet> relIsaFragment = fragmentSetOfType(IsaFragmentSet.class, fragmentSets)
+                .filter(isaFragmentSet -> isaFragmentSet.instance().equals(relation))
+                .findAny();
+
+        Optional<TypeLabel> relType = relIsaFragment
+                .map(IsaFragmentSet::type)
+                .flatMap(type -> fragmentSetOfType(LabelFragmentSet.class, fragmentSets).filter(labelFragmentSet -> labelFragmentSet.type().equals(type)).map(LabelFragmentSet::label).findAny())
+                .filter(type -> !hasDirectSubTypes(graph, type));
+
+        // Try and get role type
+        Optional<IsaCastingsFragmentSet> castingIsaFragment = fragmentSetOfType(IsaCastingsFragmentSet.class, fragmentSets)
+                .filter(isaFragmentSet -> isaFragmentSet.casting().equals(casting))
+                .findAny();
+
+        Optional<TypeLabel> roleType = castingIsaFragment
+                .map(IsaCastingsFragmentSet::roleType)
+                .flatMap(type -> fragmentSetOfType(LabelFragmentSet.class, fragmentSets).filter(labelFragmentSet -> labelFragmentSet.type().equals(type)).map(LabelFragmentSet::label).findAny());
+
+        if (castingIsaFragment.isPresent() && !roleType.isPresent()) {
+            return false;
+        }
+
+        // When the meta role-type is specified, it's the same as not specifying the role at all
+        if (roleType.isPresent() && roleType.get().equals(Schema.MetaSchema.ROLE.getLabel())) {
+            roleType = Optional.empty();
+        }
+
+        // We can't use the shortcut edge in the presence of role-type hierarchies, because we don't know precisely
```

Comment:
bummer and I am not sure how to do indexing on sets. If we could do that then the shortcut edges could contain all of the role/relation types